src/charm.py

```diff
@@ -0,0 +1,249 @@
+        It will install the exporter if not already installed or refresh the channel if the it was
+        changed.
+        """
+        self.install()
+
+        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
+        if upstream_snap.present:
+            return
+
+        snap_service = get_installed_snap_service(SNAP_NAME)
+        # if cos is not related then we should block and not run anything
+        if not self.model.relations.get("cos-agent"):
+            snap_service.stop()
+            return
+
+        snap_service.configure(
+            {
+                "cloud": CLOUD_NAME,
+                "os-client-config": str(OS_CLIENT_CONFIG),
+                "web": {"listen-address": f":{self.model.config['port']}"},
+                "cache": self.model.config["cache"],
+                "cache-ttl": self.model.config["cache_ttl"],
+            }
+        )
+
+        data = self._get_keystone_data()
+        if not data:
+            logger.info("Keystone credentials are not available, stopping services.")
+            snap_service.stop()
+            return
+
+        logger.info("Keystone credentials are available, starting services.")
+        self._write_cloud_config(data)
+        snap_service.restart_and_enable()
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install charm event."""
+        self.install()
+
+    def _on_upgrade(self, _: ops.UpgradeCharmEvent) -> None:
+        """Handle upgrade charm event."""
+        self.install()
+
+    def _on_collect_unit_status(self, event: ops.CollectStatusEvent) -> None:
+        """Handle collect unit status event (called after every event)."""
+        if not self.model.relations.get("credentials"):
+            event.add_status(BlockedStatus("Keystone is not related"))
+
+        if not self._get_keystone_data():
+            event.add_status(WaitingStatus("Waiting for credentials from keystone"))
+
+        if not self.model.relations.get("cos-agent"):
+            event.add_status(BlockedStatus("Grafana Agent is not related"))
+
+        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
+
+        # this is necessary when doing a charm upgrade coming from revision 27
+        if upstream_snap.present:
+            event.add_status(
+                BlockedStatus(
+                    "golang-openstack-exporter detected. Please see: "
+                    "https://charmhub.io/openstack-exporter#known-issues"
+                )
+            )
+
+        snap_service = get_installed_snap_service(SNAP_NAME)
+
+        if not snap_service.present:
+            event.add_status(
+                BlockedStatus("snap service is not installed, please check snap service")
+            )
+        elif not snap_service.is_active():
+            event.add_status(
+                BlockedStatus("snap service is not running, please check snap service")
+            )
```

Comment:
What would the operator do in these cases to fix the situation?

Reply:
I don't think we have mechanism to fix it right now.
Do you suggest we to auto restart the snap service on update status hook?

Reply:
The exporter is running as daemon and now is configured to restart [on-abnormal](https://github.com/canonical/charmed-openstack-exporter-snap/blob/main/snap/snapcraft.yaml#L21). I don't think the charm can do much here than blocking. Operators would need to take a look in the logs to see what is happening in my opinion.

Reply:
`BlockedStatus` implies that the operator needs to do something to resolve the issue. Perhaps include documentation for what to look for and what to do in these cases?

Reply:
There are too many potential causes for the service down, so we cannot provide the user with specific action items to resolve it. In this situation, based on the Juju framework's design philosophy, which status should be selected?  'blocked' status or 'error' status ?

Reply:
If the user can take an action to resolve/ investigate the issue, I would suggest Blocked state. If you can't come up with documentation on what the user should do, then it doesn't sound like most users would be able to resolve the issue so perhaps error state is better. Perhaps you could include some documentation on how to investigate the issue?

Reply:
Open https://github.com/canonical/openstack-exporter-operator/issues/108 to follow up this issue. We will get back here once we solve it.

Reply:
Since https://github.com/canonical/openstack-exporter-operator/pull/110 , the blocked status message is improved.  We decided to go with blocked instead of error, because this is a general status check.  And we don't want the status check to interfer with other hooks.  It's also not something that the charm will be able to resolve by retrying the hook (and the status will auto update on a future update-status hook if the issue has been resolved).

Reply:
ok, looks good